runtime/src/bank.rs

```diff
@@ -5184,12 +5184,47 @@ impl Bank {
         let count = partitions.len();
         let rent_metrics = RentMetrics::default();
         // partitions will usually be 1, but could be more if we skip slots
-        let thread_pool = &self.rc.accounts.accounts_db.thread_pool;
-        thread_pool.install(|| {
-            partitions.into_par_iter().for_each(|partition| {
+        let mut parallel = count > 1;
+        if parallel {
+            let ranges = partitions
+                .iter()
+                .map(|partition| Self::pubkey_range_from_partition(*partition))
+                .collect::<Vec<_>>();
+            // test every range to make sure ranges are not overlapping
+            // some tests collect rent from overlapping ranges
+            // example: [(0, 31, 32), (0, 0, 128), (0, 27, 128)]
+            // read-modify-write of an account for rent collection cannot be done in parallel
+            'outer: for i in 0..ranges.len() {
```

Comment:
For the case that there are no conflicts, the fully loop will actually be fully executed. O(n*n). 
Maybe we can sort the ranges first and do just a linear scan of the sorted range. 
That will reduce the complexity down to O(n*log(n)) + O(n).

Reply:
Note: once we sorted the range list, we can partition the range list into non-overlapping sublists. Those sublists themself can be parallelized if they are big enough.

Reply:
The next level of this algorithm already (as of recently) breaks up the range into non-overlapping ranges and executes them in parallel.